### PR TITLE
Add Utilities::MPI::scatter

### PR DESCRIPTION
**Pull request by mschreter**
This PR suggests to add a utility function to perform a MPI scatter operation.

@peterrum: FYI

### CODE REVIEW

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -2093,6 +2112,72 @@ namespace Utilities
 
 
 
+    template <typename T>
+    T
+    scatter(const MPI_Comm &      comm,
+            const std::vector<T> &objects_to_send,
```
Comment (mschreter) [inline include/deal.II/base/mpi.h:2118]: I was not sure if we should add an assert to check whether `objects_to_send.size()==n_procs` on `root_process`.

Comment (mschreter) [reply include/deal.II/base/mpi.h:2118]: I add an assert.

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -2093,6 +2112,76 @@ namespace Utilities
...
+            const unsigned int    root_process)
+    {
+#  ifndef DEAL_II_WITH_MPI
+      (void)comm;
+      (void)root_process;
+      std::vector<T> v(1, objects_to_send);
+      return v;
```
Comment (kronbichler) [inline include/deal.II/base/mpi.h:2125]: Shouldn't this be
```suggestion
      return  objects_to_send[0];
```

**Review by kronbichler — APPROVED** [review]
Looks good to me, apart from one minor comment for the non-MPI case.

File: include/deal.II/base/mpi.h (hunk 2 above)
Comment (mschreter) [reply include/deal.II/base/mpi.h:2125]: Thanks, absolutely. I'll fix that.

File: include/deal.II/base/mpi.h [hunk 3]
```diff
@@ -2093,6 +2112,75 @@ namespace Utilities
...
+      (void)comm;
+      (void)root_process;
+      return objects_to_send[0];
+#  else
+      const auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+      const auto my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+
+      AssertIndexRange(root_process, n_procs);
+      AssertThrow(
+        objects_to_send.size() == n_procs || my_rank != root_process,
```
Comment (peterrum) [inline include/deal.II/base/mpi.h:2131]:
```suggestion
        my_rank != root_process || objects_to_send.size() == n_procs,
```

**Review by peterrum — APPROVED** [review]
/rebuild

**Review by peterrum — APPROVED** [review]

File: include/deal.II/base/mpi.h [hunk 4]
```diff
@@ -1281,6 +1281,25 @@ namespace Utilities
...
            const unsigned int root_process = 0);
 
+    /**
+     * A generalization of the classic MPI_Scatter function, that accepts
+     * arbitrary data types T, as long as boost::serialize accepts T as an
+     * argument.
+     *
+     * @param[in] comm MPI communicator.
+     * @param[in] object_to_send a vector of objects, with size equal to the
+     * number of processes.
```
Comment (tjhei) [inline include/deal.II/base/mpi.h:1291]: Can you mention that it should be empty on all other ranks?

File: include/deal.II/base/mpi.h [hunk 5]
```diff
@@ -2093,6 +2112,75 @@ namespace Utilities
...
+      (void)comm;
+      (void)root_process;
+      return objects_to_send[0];
+#  else
+      const auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+      const auto my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+
+      AssertIndexRange(root_process, n_procs);
+      AssertThrow(
+        my_rank != root_process || objects_to_send.size() == n_procs,
```
Comment (tjhei) [inline include/deal.II/base/mpi.h:2131]: Maybe also assert that ``objects_to_send.size()==0`` on other ranks?

File: include/deal.II/base/mpi.h [hunk 6]
```diff
@@ -2093,6 +2112,75 @@ namespace Utilities
...
+                                 packed_data.end());
+              send_counts[i] = packed_data.size();
+            }
+
+          for (unsigned int i = 0; i < n_procs; ++i)
+            send_displacements[i + 1] = send_displacements[i] + send_counts[i];
+        }
+
+      int n_local_data;
+      MPI_Scatter(send_counts.data(),
```
Comment (tjhei) [inline include/deal.II/base/mpi.h:2158]: check the return value here

File: include/deal.II/base/mpi.h [hunk 7]
```diff
@@ -2093,6 +2112,75 @@ namespace Utilities
...
+                  MPI_INT,
+                  &n_local_data,
+                  1,
+                  MPI_INT,
+                  root_process,
+                  comm);
+
+      std::vector<char> recv_buffer(n_local_data);
+
+      MPI_Scatterv(send_buffer.data(),
```
Comment (tjhei) [inline include/deal.II/base/mpi.h:2169]: also check this return value

**Comment by mschreter** [comment]
@tjhei Thanks for your review, I addressed your comments in a follow-up PR #14485.

**Pull request merged**